Mathlib/Combinatorics/Enumerative/DyckWord.lean

```diff
@@ -78,46 +78,52 @@ instance : AddMonoid DyckWord where
 
 namespace DyckWord
 
-variable (p q : DyckWord)
+variable {p q : DyckWord}
 
 lemma toList_eq_nil : p.toList = [] ↔ p = 0 := by rw [DyckWord.ext_iff]; rfl
+lemma toList_ne_nil : p.toList ≠ [] ↔ p ≠ 0 := by rw [not_iff_not, toList_eq_nil]
+
+variable (h : p ≠ 0)
 
 /-- The first element of a nonempty Dyck word is `U`. -/
-lemma head_eq_U (h : p.toList ≠ []) : p.toList.head h = U := by
+lemma head_eq_U : p.toList.head (toList_ne_nil.mpr h) = U := by
   rcases p with - | s; · tauto
   rw [head_cons]
   by_contra f
   rename_i _ nonneg
   simpa [s.dichotomy.resolve_left f] using nonneg 1
 
 /-- The last element of a nonempty Dyck word is `D`. -/
-lemma getLast_eq_D (h : p.toList ≠ []) : p.toList.getLast h = D := by
+lemma getLast_eq_D : p.toList.getLast (toList_ne_nil.mpr h) = D := by
   by_contra f; have s := p.count_U_eq_count_D
-  rw [← dropLast_append_getLast h, (dichotomy _).resolve_right f] at s
+  rw [← dropLast_append_getLast (toList_ne_nil.mpr h), (dichotomy _).resolve_right f] at s
   simp_rw [dropLast_eq_take, count_append, count_singleton', ite_true, ite_false] at s
   have := p.count_D_le_count_U (p.toList.length - 1); omega
 
-lemma cons_tail_dropLast_concat (h : p.toList ≠ []) :
-    U :: p.toList.dropLast.tail ++ [D] = p := by
-  have : p.toList.dropLast.take 1 = [p.toList.head h] := by
+include h in
+lemma cons_tail_dropLast_concat : U :: p.toList.dropLast.tail ++ [D] = p := by
+  have h' := toList_ne_nil.mpr h
+  have : p.toList.dropLast.take 1 = [p.toList.head h'] := by
     rcases p with - | ⟨s, ⟨- | ⟨t, r⟩⟩⟩
     · tauto
     · rename_i bal _
       cases s <;> simp at bal
     · tauto
-  nth_rw 2 [← p.toList.dropLast_append_getLast h, ← p.toList.dropLast.take_append_drop 1]
-  rw [p.getLast_eq_D, drop_one, this, p.head_eq_U]
+  nth_rw 2 [← p.toList.dropLast_append_getLast h', ← p.toList.dropLast.take_append_drop 1]
+  rw [getLast_eq_D h, drop_one, this, head_eq_U h]
   rfl
 
+variable (p) in
 /-- Prefix of a Dyck word as a Dyck word, given that the count of `U`s and `D`s in it are equal. -/
-def take (i : ℕ) (hi : (p.toList.take i).count U = (p.toList.take i).count D) : DyckWord where
+def take {i : ℕ} (hi : (p.toList.take i).count U = (p.toList.take i).count D) : DyckWord where
```

Comment:
It's a bad idea to make `i` implicit here.
1. It's much easier to provide the proof of `(p.toList.take i).count U = (p.toList.take i).count D` if you know what `i` is
2. You don't want the user seeing `p.take ...` in the infoview (since proofs are ellipsed) and having to click through to find out what `i` is

Reply:
Yeah, I eventually deemed that as a mistake. This will be reverted in the next commit.